### ISSUE TITLE
multi gpu support for lstm layer

### ISSUE THREAD

**Issue by zachluo** [BVLC/caffe#4851]
The implement of lstm layer in caffe is to use a sub-net. But it seems to fail to use multi gpu when training lstm. It outputs: 
Check failed: Caffe::root_solver() || root_net_ root_net_ needs to be set for all non-root solvers
Any one meet this issue?

**Comment by cypof** [comment]
Does it work with the nccl version?

https://github.com/BVLC/caffe/pull/4563

**Comment by junshi15** [comment]
this PR addresses the problem. 
https://github.com/BVLC/caffe/pull/4806

**Comment by zachluo** [comment]
as @junshi15 says, it works !

**Comment by antran89** [comment]
@zachluo After apply PR #4806 of @junshi15, I have another problem. Still cannot train a LSTM model on multi-GPUs. Some memory error as following:
```
I0421 12:21:21.974892  2560 solver.cpp:280] Learning Rate Policy: multistep
F0421 12:34:01.506337  2585 syncedmem.cpp:56] Check failed: error == cudaSuccess (77 vs. 0)  an illegal memory access was encountered
*** Check failure stack trace: ***
F0421 12:34:01.506364  2586 math_functions.cu:79] Check failed: error == cudaSuccess (77 vs. 0)  an illegal memory access was encounteredF0421 12:34:01.506377  2583 benchmark.cpp:19] Check failed: error == cudaSuccess (77 vs. 0)  an illegal memory access was encountered
*** Check failure stack trace: ***
F0421 12:34:01.506364  2586 math_functions.cu:79] Check failed: error == cudaSuccess (77 vs. 0)  an illegal memory access was encounteredF0421 12:34:01.506377  2583 benchmark.cpp:19] Check failed: error == cudaSuccess (77 vs. 0)  an illegal memory access was encountered
```

**Comment by shelhamer** [comment]
Note this is solved by further releases in the switch to NCCL #4563, which is included in 1.0.
